tachyon/zk/transcript/poseidon_transcript.h

```diff
@@ -0,0 +1,129 @@
+// Copyright 2020-2022 The Electric Coin Company
+// Copyright 2022 The Halo2 developers
+// Use of this source code is governed by a MIT/Apache-2.0 style license that
+// can be found in the LICENSE-MIT.halo2 and the LICENCE-APACHE.halo2
+// file.
+
+#ifndef TACHYON_ZK_TRANSCRIPT_POSEIDON_TRANSCRIPT_H_
+#define TACHYON_ZK_TRANSCRIPT_POSEIDON_TRANSCRIPT_H_
+
+#include <utility>
+#include <vector>
+
+#include "tachyon/base/buffer/buffer.h"
+#include "tachyon/base/buffer/vector_buffer.h"
+#include "tachyon/crypto/hashes/sponge/poseidon/halo2_poseidon.h"
+#include "tachyon/zk/transcript/transcript.h"
+
+namespace tachyon::zk {
+
+template <typename AffinePointTy>
+class PoseidonRead : public TranscriptRead<AffinePointTy> {
+ public:
+  using ScalarField = typename AffinePointTy::ScalarField;
+  using Buffer = typename base::Buffer;
+
+  constexpr PoseidonRead() = default;
+  // Initialize a transcript given an input buffer.
+  explicit PoseidonRead(Buffer read_buf)
+      : state_(
+            crypto::PoseidonConfig<ScalarField>::CreateCustom(8, 5, 8, 63, 0)),
+        buffer_(std::move(read_buf)) {}
+
+  constexpr Buffer& buffer() { return buffer_; }
+
+  // TranscriptRead methods
+  constexpr bool ReadPoint(AffinePointTy* point_out) override {
+    if (buffer_.Read(point_out)) {
+      return WriteToTranscript(*point_out);
+    }
+    return false;
+  }
+
+  constexpr bool ReadScalar(ScalarField* scalar_out) override {
+    if (buffer_.Read(scalar_out)) {
+      return WriteToTranscript(*scalar_out);
+    }
+    return false;
+  }
+
+  constexpr Challenge255<AffinePointTy> SqueezeChallenge() override {
+    return Challenge255<AffinePointTy>(state_.SqueezeNativeFieldElements(1)[0]);
+  }
+
+  constexpr bool WriteToTranscript(const AffinePointTy& point) override {
+    std::vector<ScalarField> coords = {
+        ScalarField::FromBigInt(point.x().ToBigInt() %
+                                ScalarField::Config::kModulus),
+        ScalarField::FromBigInt(point.y().ToBigInt() %
+                                ScalarField::Config::kModulus)};
+    return state_.Absorb(coords);
+  }
+
+  constexpr bool WriteToTranscript(const ScalarField& scalar) override {
+    return state_.Absorb(scalar);
+  }
+
+ private:
+  crypto::Halo2PoseidonSponge<ScalarField> state_;
+  Buffer buffer_;
+};
+
+template <typename AffinePointTy>
+class PoseidonWrite : public TranscriptWrite<AffinePointTy> {
+ public:
+  using ScalarField = typename AffinePointTy::ScalarField;
+  using VectorBuffer = base::VectorBuffer;
```

Comment:
If it is this much short, i think it's better to use a plain `base::Buffer`.

Reply:
I fixed to use `base::VectorBuffer`!